File: src/backend/services/analytics/src/config/elasticsearch.py
# External dependencies
# elasticsearch v8.0.0 - Official Elasticsearch client
from elasticsearch import AsyncElasticsearch, RequestsHttpConnection
from elasticsearch.exceptions import ConnectionError, TransportError
# elasticsearch_dsl v8.0.0 - High-level Elasticsearch DSL
from elasticsearch_dsl import Index, IndexTemplate
# Standard library imports
from typing import Dict, Optional, Any, List
import os
import asyncio
from datetime import datetime

# Internal imports
from ../../shared/utils/logger import Logger

# Global configuration with secure defaults and environment variable overrides
ELASTICSEARCH_CONFIG = {
    'hosts': os.getenv('ELASTICSEARCH_HOSTS', 'localhost:9200').split(','),
    'username': os.getenv('ELASTICSEARCH_USERNAME', 'elastic'),
    'password': os.getenv('ELASTICSEARCH_PASSWORD'),
    'use_ssl': os.getenv('ELASTICSEARCH_USE_SSL', 'false').lower() == 'true',
    'verify_certs': os.getenv('ELASTICSEARCH_VERIFY_CERTS', 'true').lower() == 'true',
    'retry_on_timeout': True,
    'max_retries': 3,
    'timeout': 30,
    'maxsize': int(os.getenv('ELASTICSEARCH_POOL_SIZE', 25)),
    'sniff_on_start': True,
    'sniff_on_connection_fail': True,
    'sniffer_timeout': 60,
    'http_compress': True,
    'connection_class': RequestsHttpConnection
}

# Index settings optimized for analytics workload
INDEX_SETTINGS = {
    'number_of_shards': int(os.getenv('ES_NUMBER_OF_SHARDS', 3)),
    'number_of_replicas': int(os.getenv('ES_NUMBER_OF_REPLICAS', 1)),
    'refresh_interval': os.getenv('ES_REFRESH_INTERVAL', '1s'),
    'analysis': {
        'analyzer': {
            'custom_analyzer': {
                'type': 'custom',
                'tokenizer': 'standard',
                'filter': ['lowercase', 'stop', 'snowball']
            }
        }
    },
    'index.lifecycle.name': 'analytics_policy',
    'index.routing.allocation.total_shards_per_node': 3
}

class ElasticsearchManager:
    """
    Manages Elasticsearch connections and operations for analytics service with 
    advanced cluster management and monitoring capabilities.
    """
    
    def __init__(self):
        """Initialize Elasticsearch manager with enhanced configuration."""
        self._client: Optional[AsyncElasticsearch] = None
        self._logger = Logger("elasticsearch_manager")
        self._config = ELASTICSEARCH_CONFIG.copy()
        self._health_status = {
            'last_check': None,
            'status': None,
            'node_count': 0,
            'shards': {},
            'indices': {}
        }

    async def initialize(self) -> None:
        """
        Initialize Elasticsearch client and verify connection with enhanced error handling.
        """
        try:
            # Create client instance with retry logic
            self._client = AsyncElasticsearch(**self._config)
            
            # Verify cluster connection
            await self._client.ping()
            
            # Check initial cluster health
            health = await self._client.cluster.health()
            self._health_status.update({
                'last_check': datetime.utcnow(),
                'status': health['status'],
                'node_count': health['number_of_nodes']
            })
            
            self._logger.info("Elasticsearch client initialized successfully", {
                'cluster_name': health['cluster_name'],
                'node_count': health['number_of_nodes'],
                'status': health['status']
            })
            
        except ConnectionError as e:
            self._logger.error("Failed to connect to Elasticsearch cluster", e, {
                'hosts': self._config['hosts']
            })
            raise
        except Exception as e:
            self._logger.error("Unexpected error during Elasticsearch initialization", e)
            raise

    async def check_health(self) -> Dict[str, Any]:
        """
        Performs comprehensive health check of Elasticsearch cluster.
        
        Returns:
            Dict containing detailed health metrics and status information.
        """
        if not self._client:
            raise RuntimeError("Elasticsearch client not initialized")
            
        try:
            # Get cluster health
            health = await self._client.cluster.health()
            
            # Get node stats
            nodes = await self._client.nodes.stats()
            
            # Get indices health
            indices = await self._client.cat.indices(format='json')
            
            # Update internal health tracking
            self._health_status.update({
                'last_check': datetime.utcnow(),
                'status': health['status'],
                'node_count': len(nodes['nodes']),
                'shards': {
                    'total': health['active_shards'],
                    'relocating': health['relocating_shards'],
                    'unassigned': health['unassigned_shards']
                },
                'indices': {idx['index']: idx['health'] for idx in indices}
            })
            
            return self._health_status
            
        except Exception as e:
            self._logger.error("Error checking Elasticsearch health", e)
            raise

    async def create_index_template(self, template_name: str, template_body: Dict[str, Any]) -> bool:
        """
        Creates or updates index template with specified settings.
        
        Args:
            template_name: Name of the template to create/update
            template_body: Template configuration and mappings
            
        Returns:
            bool: Success status of template creation
        """
        if not self._client:
            raise RuntimeError("Elasticsearch client not initialized")
            
        try:
            # Merge default settings with template settings
            template_body['settings'] = {
                **INDEX_SETTINGS,
                **template_body.get('settings', {})
            }
            
            # Create/update template
            response = await self._client.indices.put_template(
                name=template_name,
                body=template_body
            )
            
            success = response.get('acknowledged', False)
            if success:
                self._logger.info(f"Successfully created index template: {template_name}")
            else:
                self._logger.warn(f"Index template creation not acknowledged: {template_name}")
                
            return success
            
        except Exception as e:
            self._logger.error(f"Error creating index template: {template_name}", e)
            raise

async def create_elasticsearch_manager() -> ElasticsearchManager:
    """
    Factory function to create ElasticsearchManager instance with enhanced configuration.
    
    Returns:
        Initialized ElasticsearchManager instance
    """
    manager = ElasticsearchManager()
    await manager.initialize()
    await manager.check_health()
    return manager

File: src/backend/services/candidate/src/config/elasticsearch.ts
// External dependencies
// @elastic/elasticsearch v8.0.0 - Official Elasticsearch client
import { Client, ConfigOptions } from '@elastic/elasticsearch';
import { errors } from '@elastic/elasticsearch';

// Internal dependencies
import { Logger } from '../../shared/utils/logger';

// Environment-based configuration with secure defaults
const ES_NODE = process.env.ELASTICSEARCH_NODE || 'http://localhost:9200';
const ES_USERNAME = process.env.ELASTICSEARCH_USERNAME;
const ES_PASSWORD = process.env.ELASTICSEARCH_PASSWORD;
const ES_INDEX_PREFIX = 'refactortrack_candidates';
const ES_MAX_RETRIES = Number(process.env.ELASTICSEARCH_MAX_RETRIES || 3);
const ES_REQUEST_TIMEOUT = Number(process.env.ELASTICSEARCH_REQUEST_TIMEOUT || 30000);
const ES_PING_TIMEOUT = Number(process.env.ELASTICSEARCH_PING_TIMEOUT || 3000);
const ES_SNIFF_ON_START = process.env.ELASTICSEARCH_SNIFF_ON_START === 'true';
const ES_SSL_VERIFY = process.env.NODE_ENV === 'production';

/**
 * Enhanced configuration class for Elasticsearch with comprehensive error handling
 * and health monitoring capabilities.
 */
export class ElasticsearchConfig {
  private client: Client;
  private logger: Logger;
  private options: ConfigOptions;
  private isConnected: boolean = false;
  private healthCheckInterval: NodeJS.Timer | null = null;

  constructor(options: ConfigOptions = {}) {
    this.logger = new Logger('elasticsearch-config', {
      additionalMeta: { service: 'candidate-service' }
    });

    this.options = this.mergeConfig(options);
    this.client = this.initializeClient();
    this.setupHealthCheck();
  }

  /**
   * Merges provided configuration with default settings
   */
  private mergeConfig(options: ConfigOptions): ConfigOptions {
    const defaultConfig: ConfigOptions = {
      node: ES_NODE,
      auth: ES_USERNAME && ES_PASSWORD ? {
        username: ES_USERNAME,
        password: ES_PASSWORD
      } : undefined,
      maxRetries: ES_MAX_RETRIES,
      requestTimeout: ES_REQUEST_TIMEOUT,
      pingTimeout: ES_PING_TIMEOUT,
      sniffOnStart: ES_SNIFF_ON_START,
      ssl: ES_SSL_VERIFY ? {
        rejectUnauthorized: true
      } : undefined
    };

    return { ...defaultConfig, ...options };
  }

  /**
   * Initializes Elasticsearch client with error handling
   */
  private initializeClient(): Client {
    try {
      const client = new Client(this.options);
      this.logger.info('Elasticsearch client initialized', { node: ES_NODE });
      return client;
    } catch (error) {
      this.logger.error('Failed to initialize Elasticsearch client', error as Error);
      throw error;
    }
  }

  /**
   * Sets up periodic health checks for the Elasticsearch connection
   */
  private setupHealthCheck(): void {
    this.healthCheckInterval = setInterval(async () => {
      try {
        await this.checkConnection();
      } catch (error) {
        this.logger.error('Health check failed', error as Error);
        this.isConnected = false;
      }
    }, ES_PING_TIMEOUT);
  }

  /**
   * Returns validated Elasticsearch client instance
   */
  public async getClient(): Promise<Client> {
    if (!this.isConnected) {
      await this.checkConnection();
    }
    return this.client;
  }

  /**
   * Comprehensive connection and cluster health verification
   */
  public async checkConnection(): Promise<boolean> {
    try {
      // Ping test
      await this.client.ping({ timeout: ES_PING_TIMEOUT });

      // Cluster health check
      const health = await this.client.cluster.health();
      
      if (health.status === 'red') {
        this.logger.warn('Elasticsearch cluster health is red', { health });
        return false;
      }

      this.isConnected = true;
      this.logger.debug('Elasticsearch connection verified', { 
        health,
        node: ES_NODE 
      });
      
      return true;
    } catch (error) {
      this.isConnected = false;
      if (error instanceof errors.ConnectionError) {
        this.logger.error('Elasticsearch connection error', error);
      } else if (error instanceof errors.TimeoutError) {
        this.logger.error('Elasticsearch timeout error', error);
      } else {
        this.logger.error('Elasticsearch check failed', error as Error);
      }
      throw error;
    }
  }

  /**
   * Creates and maintains required indices with proper mappings
   */
  public async createIndices(): Promise<void> {
    try {
      const indexName = getIndexName('candidate', process.env.NODE_ENV || 'development');
      
      const indexExists = await this.client.indices.exists({ index: indexName });
      
      if (!indexExists) {
        await this.client.indices.create({
          index: indexName,
          body: {
            settings: {
              number_of_shards: 3,
              number_of_replicas: 2,
              analysis: {
                analyzer: {
                  resume_analyzer: {
                    type: 'custom',
                    tokenizer: 'standard',
                    filter: ['lowercase', 'stop', 'snowball']
                  }
                }
              }
            },
            mappings: {
              properties: {
                id: { type: 'keyword' },
                firstName: { type: 'text' },
                lastName: { type: 'text' },
                email: { type: 'keyword' },
                skills: { type: 'keyword' },
                resumeText: { 
                  type: 'text',
                  analyzer: 'resume_analyzer'
                },
                createdAt: { type: 'date' },
                updatedAt: { type: 'date' }
              }
            }
          }
        });
        
        this.logger.info('Created Elasticsearch index', { index: indexName });
      }
    } catch (error) {
      this.logger.error('Failed to create indices', error as Error);
      throw error;
    }
  }

  /**
   * Cleanup resources on service shutdown
   */
  public async cleanup(): Promise<void> {
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
    }
    await this.client.close();
    this.logger.info('Elasticsearch resources cleaned up');
  }
}

/**
 * Creates and configures an Elasticsearch client instance
 */
export async function createElasticsearchClient(options: ConfigOptions = {}): Promise<Client> {
  const config = new ElasticsearchConfig(options);
  await config.checkConnection();
  await config.createIndices();
  return config.getClient();
}

/**
 * Generates environment-aware index names
 */
export function getIndexName(entityType: string, environment: string): string {
  if (!entityType) {
    throw new Error('Entity type is required for index name generation');
  }
  
  const envSuffix = environment !== 'production' ? `_${environment}` : '';
  return `${ES_INDEX_PREFIX}_${entityType}${envSuffix}`.toLowerCase();
}